lib/src/frontend/stream_matchers.dart

```diff
@@ -0,0 +1,372 @@
+        } catch (error, stackTrace) {
+          if (error != null) {
+            firstError = error;
+            firstStackTrace = stackTrace;
+          }
+          return;
+        }
+
+        if (result != null) {
+          failures[i] = result;
+        } else if (consumedMost == null ||
+            consumedMost.eventsDispatched < copy.eventsDispatched) {
+          consumedMost = copy;
+        }
+      }());
+    }
+
+    await Future.wait(futures);
+
+    if (consumedMost == null) {
+      transaction.reject();
+      if (firstError != null) {
+        await new Future.error(firstError, firstStackTrace);
+      }
+
+      var failureMessages = <String>[];
+      for (var i = 0; i < matchers.length; i++) {
+        var message = "failed to ${streamMatchers[i].description}";
+        if (failures[i].isNotEmpty) {
+          message += message.contains("\n") ? "\n" : " ";
+          message += "because it ${failures[i]}";
+        }
+
+        failureMessages.add(message);
+      }
+
+      return "failed all options:\n${bullet(failureMessages)}";
+    } else {
+      transaction.commit(consumedMost);
+      return null;
+    }
+  }, description);
+}
+
+/// Returns a [StreamMatcher] that matches the stream if each matcher in
+/// [matchers] matches, one after another.
+///
+/// If any matcher fails to match, this fails and consumes no events.
+StreamMatcher emitsInOrder(Iterable matchers) {
+  var streamMatchers = matchers.map(emits).toList();
+  if (streamMatchers.length == 1) return streamMatchers.first;
+
+  var description = "do the following in order:\n" + bullet(
+      streamMatchers.map((matcher) => matcher.description));
+
+  return new StreamMatcher((queue) async {
+    for (var i = 0; i < streamMatchers.length; i++) {
+      var matcher = streamMatchers[i];
+      var result = await matcher.matchQueue(queue);
+      if (result == null) continue;
+
+      var newResult = "didn't ${matcher.description}";
+      if (result.isNotEmpty) {
+        newResult += newResult.contains("\n") ? "\n" : " ";
+        newResult += "because it $result";
+      }
+      return newResult;
+    }
+  }, description);
+}
+
+/// Returns a [StreamMatcher] that matches any number of
+/// events followed by events that match [matcher].
+///
+/// This consumes all events matched by [matcher], as well as all events before.
+/// If the stream emits a done event without matching [matcher], this fails and
+/// consumes no events.
+StreamMatcher emitsThrough(matcher) {
+  var streamMatcher = emits(matcher);
+  return new StreamMatcher((queue) async {
+    var failures = <String>[];
+
+    tryHere() => queue.withTransaction((copy) async {
+      var result = await streamMatcher.matchQueue(copy);
+      if (result == null) return true;
+      failures.add(result);
+      return false;
+    });
+
+    while (await queue.hasNext) {
+      if (await tryHere()) return null;
+      await queue.next;
+    }
```

Comment:
This function is quadratic, right? Or I guess, more specifically O(mn) where m is the stream length and n is the matcher's length.

That's OK, but it would probably be good to mention it since it can get slow with long streams and matchers.

Or you could adapt Boyer-Moore to this. 😛